### PR TITLE
SslHandler avoid calling wrap/unwrap when unnecessary

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
The SSLEngine wrap and unwrap methods can be called in a way that has no side effects, but this could involve costly validation and allocation. The SslHandler should avoid calling into these methods if possible.

Modifications:
- wrapNonAppData should provide additional status which can be used by wrap to breakout early if possible

Result:
SslHandler invokes the SSLEngine less.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@Scottmitch please check unit test failures.

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java [hunk 1]
```diff
@@ -1113,6 +1117,7 @@ private boolean unwrap(
         try {
             // Only continue to loop if the handler was not removed in the meantime.
             // See https://github.com/netty/netty/issues/5860
+            unwrapLoop:
             while (!ctx.isRemoved()) {
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: @Scottmitch can you put this in front of the while just to make it more clear:

```
unwrapLoop: while(....)
```

Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: sure done

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java [hunk 2]
```diff
@@ -725,8 +725,11 @@ private void finishWrap(ChannelHandlerContext ctx, ByteBuf out, ChannelPromise p
         }
     }
 
-    // This method will not call setHandshakeFailure(...) !
-    private void wrapNonAppData(ChannelHandlerContext ctx, boolean inUnwrap) throws SSLException {
+    /**
+     * This method will not call setHandshakeFailure(...) !
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: @Scottmitch as you changed this to java docs please use `{@link #setHandshakeFailure(.....`

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java [hunk 3]
```diff
@@ -1163,7 +1167,9 @@ private boolean unwrap(
                     case NEED_UNWRAP:
                         break;
                     case NEED_WRAP:
-                        wrapNonAppData(ctx, true);
+                        if (wrapNonAppData(ctx, true) && length == 0) {
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: can you add some comments why its ok to break the loop in this case ?

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java [hunk 4]
```diff
@@ -1196,7 +1202,9 @@ private boolean unwrap(
                             flushedBeforeHandshake = false;
                             wrapLater = true;
                         }
-
+                        if (length == 0) {
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: can you add some comments why its ok to break the loop in this case ?

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java (hunk 2 above)
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: done

**Comment by normanmaurer** [comment]
@Scottmitch ship it

**Comment by Scottmitch** [comment]
4.1 (6353c22) 4.0 (ab65b1e)

**Pull request closed without merging**